src/daemon/daemon.cpp

```diff
@@ -2133,6 +2133,11 @@ try // clang-format on
         std::function<grpc::Status(VirtualMachine&)> operation;
         if (request->cancel_shutdown())
             operation = std::bind(&Daemon::cancel_vm_shutdown, this, std::placeholders::_1);
+        else if (request->force_stop())
+        {
+            auto adapted_switch_off_vm = [this](auto&& arg) -> grpc::Status { return this->switch_off_vm(arg); };
```

Comment:
Well, both `std::bind` and lambda adapt the function interface. I would say lambda is definitely a preferred option due to C++ modernity, readability, and performance. 

About performance, https://www.youtube.com/watch?v=uzF4u9KgUWI&t=3121s 45:35 also gave some benchmark results on the compile time and run time cost of `std::bind`. 

About the universal reference parameter. Well, that `(auto&& arg)` is just a general idiom for using lambda as an adapter.

Reply:
In item 34 of Scott Meyers's book, the argument to prefer lambdas to `std::bind` rests largely on lambdas being "more readable, more expressive, and may be more efficient than using `std::bind`". The argument in that talk is also mainly efficiency (compile-time and runtime) + the appeal to Scott Meyers's authority.

IMO, "more readable" and "more expressive" is unfair in many cases. This is one such example: the bind version is shorter and more obvious. Errata comment "6/15/14 tyk All      Item 34" expresses that idea too. I don't think there was a second edition of this book, but the fact that he mentioned this in there is indication that he would probably express a different view today. When I read the book I had this errata already and so I always took this item with a grain of salt.

As for efficiency, 15% overhead to compile this line plus twice the function-call time is entirely negligible in here. Definitely a price I would pay for the more obvious code.

Note that I am not advocating using bind instead of lambdas everywhere. At all, I default to lambdas myself. But in some situations, `std::bind` is clearer.

Reply:
> About the universal reference parameter. Well, that (auto&& arg) is just a general idiom for using lambda as an adapter. I think it signals better that lambda is being used as an adapter.

I don't think that general idiom makes sense without a `std::forward`. When you consider adding that, along with consistency with the other if-branches, it becomes clear which one is more readable/expressive. 

I mean, just have a look at the alternatives:

```cpp
    if (status.ok())
    {
        assert(instance_selection.deleted_selection.empty());
        assert(instance_selection.missing_instances.empty());

        std::function<grpc::Status(VirtualMachine&)> operation;
        if (request->cancel_shutdown())
            operation = std::bind(&Daemon::cancel_vm_shutdown, this, std::placeholders::_1);
        else if (request->force_stop())
            operation = std::bind(&Daemon::adapted_switch_off_vm, this, std::placeholders::_1);
        else
            operation = std::bind(&Daemon::shutdown_vm, this, std::placeholders::_1,
                                  std::chrono::minutes(request->time_minutes()));

        status = cmd_vms(instance_selection.operative_selection, operation);
    }
```

or 

```cpp
    if (status.ok())
    {
        assert(instance_selection.deleted_selection.empty());
        assert(instance_selection.missing_instances.empty());

        std::function<grpc::Status(VirtualMachine&)> operation;
        if (request->cancel_shutdown())
            operation = std::bind(&Daemon::cancel_vm_shutdown, this, std::placeholders::_1);
        else if (request->force_stop())
        {
            auto adapted_switch_off_vm = [this](auto&& arg) -> grpc::Status {
                return this->switch_off_vm(std::forward<decltype(arg)>(arg)); 
            };
            operation = adapted_switch_off_vm;
        }
        else
            operation = std::bind(&Daemon::shutdown_vm, this, std::placeholders::_1,
                                  std::chrono::minutes(request->time_minutes()));

        status = cmd_vms(instance_selection.operative_selection, operation);
    }
```

Which of those is clearer?

Reply:
One concern of mine is that the run-time and compile-time cost might add up if we continue to use `std::bind`. My opinion here is that the new one should start with lambda. 

You mentioned the verboseness of the lambda syntax, it can be compressed into one-liner easily              `operation = [this](VirtualMachine& vm) { return this->switch_off_vm(vm); };` by removing the trailing return, functor name, and converting the universal reference to the exact parameter `switch_off_vm` needs. 

You also touched upon the consistency of the other two branches, then I think we'd better change the three of them.

Reply:
All in all, I think a fair comparison should be 
```
        std::function<grpc::Status(VirtualMachine&)> operation;
        if (request->cancel_shutdown())
            operation = std::bind(&Daemon::cancel_vm_shutdown, this, std::placeholders::_1);
        else if (request->force_stop())
            operation = std::bind(&Daemon::switch_off_vm, this, std::placeholders::_1);
        else
            operation = std::bind(&Daemon::shutdown_vm, this, std::placeholders::_1,
                                  std::chrono::minutes(request->time_minutes()));
```
vs
```
        std::function<grpc::Status(VirtualMachine&)> operation;
        if (request->cancel_shutdown())
            operation = [this](const VirtualMachine& vm) { return this->cancel_vm_shutdown(vm); };
        else if (request->force_stop())
            operation = [this](VirtualMachine& vm) { return this->switch_off_vm(vm); };
        else
            operation = [this, delay_minutes = std::chrono::minutes(request->time_minutes())](VirtualMachine& vm) {
                return this->shutdown_vm(vm, delay_minutes);
            };
```

Reply:
Yes, I agree that lambdas should be the default tool. Repeating what I said above *I default to lambdas myself*. It is just that in some cases, bind still makes it more obvious. And I wonder if the penalties are still the same with today's compilers as they were 8+ years ago. Even then, it is probably a matter of an extra pointer dereference...

But OK, the second version you have above is fine. I still find the first one marginally more obvious, personally, but I am perfectly fine with the other one. Let's change all of them then.

Reply:
>  I still find the first one marginally more obvious.
I kind of agree that `std::bind` does make the adapter thing clear if you are familiar with the `std::placeholders` syntax. 

> And I wonder if the penalties are still the same with today's compilers as they were 8+ years ago
It is likely that std::bind must handle various types of argument forwarding and handle placeholders dynamically. The actually runtime cost penalty probably is negligible. Compile-time wise, lambda does not have TMP involved, so there will be some gain on this.